system/hiera.py

```diff
@@ -0,0 +1,200 @@
+      copy: src=/var/lib/hiera/ dest=/var/lib/
+
+    - name: Copy Hiera Config into Destination node
+      copy: src=/etc/hiera.yaml dest=/etc/hiera.yaml
+
+    - name: Retrieving Hiera Data
+      hiera: path=/bin/hiera key="{{ item.value }}" fact="{{ item.key }}" source=/etc/hiera.yaml
+      args:
+        context:
+          environment: 'production'
+          fqdn: 'puppet01.localdomain'
+      with_dict:
+        var_array_multi: "proxy::array_multi"
+        var_array_line: "proxy::array_line"
+        line: "line"
+
+    - debug: msg="{{ item }}"
+      with_items: var_array_multi
+    - debug: msg="{{ item }}"
+      with_items: var_array_line
+    - debug: msg="{{ line }}"
+'''
+
+RETURN = '''
+path:
+    description: bin for hiera execution
+    returned: success
+    type: string
+    sample: "/path/to/file"
+environment:
+    description: scope to be interpreted by hiera
+    returned: success even the scope does not exists
+    type: string
+    sample: "production"
+fqdn:
+    description: other scope for hiera parse
+    returned: success even the scope does not exists
+    type: string
+    sample: ""
+key:
+    description: hiera's value name
+    returned: success
+    type: string
+    sample: "proxy::array_multi"
+fact:
+    description: variable where store the Hiera's value
+    returned: success
+    type: string
+    sample: "var_array_multi"
+source:
+    description: hiera config file
+    returned: success
+    type: string
+    sample: "/path/to/hiera.yaml"
+'''
+
+
+def main():
+    """
+    Main function.
+    This module will parse your hiera hierarchi and will return the required
+    values
+    - key: Is the key name of the hiera variable
+    - fact: Is the key name that must store the hiera output
+    - args: context: must contain all the values that identify the node against hiera
+    - path: hiera executable
+    - source: hiera config file
+        WARNING: This module try to solve a disparity between arch of
+        Puppet/Hiera and Ansible, because of how they works (Puppet compile
+        the values into puppet master node) (Ansible are executed into
+        destination node and have not access to Hiera backend directly)
+
+    """
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(required=True, aliases=['key']),
+            fact=dict(required=False),
+            path=dict(required=False, default="hiera"),
+            context=dict(required=False, default={}, type='dict'),
+            source=dict(required=False, default=None)
+        )
+    )
+
+    params = module.params
+    out = {}
+
+    if not params['fact']:
+        params['fact'] = params['name']
+
+    try:
+        pargs = [params['path']]
+
+        if params['source']:
+            pargs.extend(['-c', params['source']])
+
+        pargs.append(params['name'])
+        pargs.extend(
+            [r'%s=%s' % (k, v) for k, v in params['context'].iteritems()]
```

Comment:
In order to get hiera to 'hierarchy' you need to provide filtering variables.

environment=production fqdn=host1.example.com osfamily=Debian etc etc

I think you'd have  to unfold all of facts['FACTER_*'] into key=value pairs and append those to the end of the command to get it to work. Otherwise you're just looking things up in a single namespace.